backends/fs/android/android-fs-factory.cpp

```diff
@@ -0,0 +1,100 @@
+/* ScummVM - Graphic Adventure Engine
+ *
+ * ScummVM is the legal property of its developers, whose names
+ * are too numerous to list here. Please refer to the COPYRIGHT
+ * file distributed with this source distribution.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#if defined(__ANDROID__)
+
+#include "backends/platform/android/jni-android.h"
+
+#include "backends/fs/android/android-fs-factory.h"
+#include "backends/fs/android/android-posix-fs.h"
+#include "backends/fs/android/android-saf-fs.h"
+
+namespace Common {
+DECLARE_SINGLETON(AndroidFilesystemFactory);
+}
+
+AndroidFilesystemFactory::AndroidFilesystemFactory() : _withSAF(false) {
+	AndroidSAFFilesystemNode::initJNI();
+}
+
+void AndroidFilesystemFactory::initSAF() {
+	_withSAF = true;
+	AndroidSAFFilesystemNode::initJNI();
+}
+AbstractFSNode *AndroidFilesystemFactory::makeRootFileNode() const {
+	return new AndroidPOSIXFilesystemNode(_config);
+}
+
+AbstractFSNode *AndroidFilesystemFactory::makeCurrentDirectoryFileNode() const {
+	// As current working directory can point outside of our data don't take any risk
+	return makeRootFileNode();
+}
+
+AbstractFSNode *AndroidFilesystemFactory::makeFileNodePath(const Common::String &path) const {
+	if (path.empty() || path.equals("/")) {
+		return makeRootFileNode();
+	}
+
+	// No need to take SAF add mode here as it's called only for paths and we won't accept /saf path to make a new SAF
+
+	// If SAF works, it was a SAF URL
+	AbstractFSNode *node = AndroidSAFFilesystemNode::makeFromPath(path);
+	if (node) {
+		return node;
+	}
+
+	return new AndroidPOSIXFilesystemNode(path, _config);
+}
+
+void AndroidFilesystemFactory::updateDrives() {
+	_config.drives.clear();
+	_driveNames.clear();
+
+	Common::Array<jobject> trees = JNI::getSAFTrees();
+	Common::Array<Common::String> list = JNI::getAllStorageLocations();
+	_config.drives.reserve(trees.size() + list.size() / 2);
+	_driveNames.reserve(trees.size() + list.size() / 2);
+
+	// For SAF
+	for (Common::Array<jobject>::iterator it = trees.begin(); it != trees.end(); it++) {
+		Common::String displayName;
+		Common::String path = AndroidSAFFilesystemNode::getTreePath(*it, displayName);
+		if (path.empty()) {
+			continue;
+		}
+		_driveNames.push_back(displayName);
+		_config.drives.push_back(path);
+	}
+
+	// For old POSIX way
```

Comment:
This code branch runs even if `_withSAF` is false. Is this intentional?

Reply:
Thanks, I added _withSAF on the late and missed a few spots.
SAF part should not be iterated as JNI parameters are not initialized.
POSIX part must run as it's the legacy drives I removed from `PosixFilesystemNode`